### PR TITLE
Support commonprefix in listObjects to match s3cmd ls behavior

### PR DESCRIPTION
**Pull request by shuzhang1989**

### CODE REVIEW

File: common/s3util.cpp [hunk 1]
```diff
@@ -162,17 +162,29 @@ SdkGetObjectResponse S3Util::sdkGetObject(const string& key,
...
+  }
   auto listObjectResult = s3Client.ListObjects(listObjectRequest);
   if (listObjectResult.IsSuccess()) {
-    Aws::Vector<Aws::S3::Model::Object> contents =
-      listObjectResult.GetResult().GetContents();
-    for (auto object : contents) {
-      objects.push_back(object.GetKey());
+    if (!delimiter.empty()) {
+      Aws::Vector<Aws::S3::Model::CommonPrefix> contents =
+      listObjectResult.GetResult().GetCommonPrefixes();
```
Comment (angxu) [inline common/s3util.cpp]: nit: indent

Comment (shuzhang1989) [inline common/s3util.cpp]: ❤️

File: common/s3util.cpp [hunk 2]
```diff
@@ -162,17 +162,29 @@ SdkGetObjectResponse S3Util::sdkGetObject(const string& key,
...
 
-ListObjectsResponse S3Util::listObjects(const string& prefix) {
+ListObjectsResponse S3Util::listObjects(const string& prefix,
+                                        const string& delimiter) {
   vector<string> objects;
   ListObjectsRequest listObjectRequest;
   listObjectRequest.SetBucket(bucket_);
   listObjectRequest.SetPrefix(prefix);
+  if (!delimiter.empty()) {
+    listObjectRequest.SetDelimiter("/");
```
Comment (hrjaco) [inline common/s3util.cpp]: replace "/" with delimiter?

File: common/s3util.cpp [hunk 3]
```diff
@@ -162,17 +162,29 @@ SdkGetObjectResponse S3Util::sdkGetObject(const string& key,
...
   auto listObjectResult = s3Client.ListObjects(listObjectRequest);
   if (listObjectResult.IsSuccess()) {
-    Aws::Vector<Aws::S3::Model::Object> contents =
-      listObjectResult.GetResult().GetContents();
-    for (auto object : contents) {
-      objects.push_back(object.GetKey());
+    if (!delimiter.empty()) {
+      Aws::Vector<Aws::S3::Model::CommonPrefix> contents =
+        listObjectResult.GetResult().GetCommonPrefixes();
+      for (auto object : contents) {
```
Comment (hrjaco) [inline common/s3util.cpp]: Change to const auto&?

File: common/s3util.cpp [hunk 4]
```diff
@@ -162,17 +162,29 @@ SdkGetObjectResponse S3Util::sdkGetObject(const string& key,
...
+        objects.push_back(object.GetPrefix());
+      }
+    } else {
+      Aws::Vector<Aws::S3::Model::Object> contents =
+        listObjectResult.GetResult().GetContents();
+      for (auto object : contents) {
+        objects.push_back(object.GetKey());
```
Comment (hrjaco) [inline common/s3util.cpp]: Same here.

File: common/s3util.cpp (hunk 2 above)
Comment (shuzhang1989) [inline common/s3util.cpp]: yes, fixed

**Comment by hrjaco** [comment]
thanks a lot Shu!! r+

**Pull request merged**